### PR TITLE
APEXMALHAR-2205 #resolve #comment State management benchmark - add up…

### PR DESCRIPTION
**Pull request by brightchen**
…date

### CODE REVIEW

**Comment by brightchen** [comment]
@tweise Please review. Added update

File: benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java [hunk 1]
```diff
@@ -34,6 +41,13 @@
 {
   private static final Logger logger = LoggerFactory.getLogger(StoreOperator.class);
 
+  public static enum ExeMode
```
Comment (tweise) [inline benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java]: ExecMode?

File: benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java [hunk 2]
```diff
@@ -34,6 +41,13 @@
...
+  {
+    Insert,
```
Comment (tweise) [inline benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java]: All uppercase?

File: benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java [hunk 3]
```diff
@@ -80,10 +97,61 @@ public void endWindow()
...
+  protected transient Map<Future<Slice>, KeyValPair<byte[], byte[]>> taskToPair = Maps.newHashMap();
+  
   protected void processTuple(KeyValPair<byte[], byte[]> tuple)
   {
+    if (ExeMode.UpdateAsync == exeMode) {
+      updateAsync(tuple);
+      return;
+    }
+
+    Slice key = new Slice(tuple.getKey());
```
Comment (tweise) [inline benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java]: move into if block

File: benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java [hunk 4]
```diff
@@ -80,10 +97,61 @@ public void endWindow()
     }
   }
 
+  protected transient Queue<Future<Slice>> taskQueue = new LinkedList<Future<Slice>>();
+  protected transient Map<Future<Slice>, KeyValPair<byte[], byte[]>> taskToPair = Maps.newHashMap();
+  
   protected void processTuple(KeyValPair<byte[], byte[]> tuple)
   {
+    if (ExeMode.UpdateAsync == exeMode) {
```
Comment (tweise) [inline benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java]: Shouldn't this be a case statement, with insertValueToStore as default?

File: benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java [hunk 5]
```diff
@@ -80,10 +97,61 @@ public void endWindow()
...
+    Slice key = new Slice(tuple.getKey());
+
+    if (ExeMode.UpdateSync == exeMode) {
+      store.getSync(bucketId, key);
+    }
+    insertValueToStore(tuple);
+  }
+  
+  protected final int taskBarrier = 1000000;
+  protected void updateAsync(KeyValPair<byte[], byte[]> tuple)
```
Comment (tweise) [inline benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java:141]: below is blocking, why is it called "async"?

File: benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java [hunk 6]
```diff
@@ -80,10 +97,61 @@ public void endWindow()
...
+        logger.info("Queue Size: {}, wait time(milli-seconds): {}", taskQueue.size(), taskQueue.size() / taskBarrier);
+        Thread.sleep(taskQueue.size() / taskBarrier);
+      } catch (Exception e) {
+        //ignore
+      }
+    }
+    
+    {
+      Slice key = new Slice(tuple.getKey());
+      Future<Slice> task = store.getAsync(bucketId, key);
```
Comment (tweise) [inline benchmark/src/main/java/com/datatorrent/benchmark/state/StoreOperator.java]: Why do an async get when the required behavior is blocking?

**Review by tweise — CHANGES_REQUESTED** [review]

**Comment by brightchen** [comment]
@tweise please review. The code already changed, but the code in conversion page is still the old one

**Pull request merged**